src/routing/effects/index.ts

```diff
@@ -0,0 +1,372 @@
+  const replace = data.data.replace || historyChange === 'replace';
+
+  if (historyChange !== 'back' && !equals(location.path, newPath)) {
+    history[replace ? 'replaceState' : 'pushState']({ state: JSON.stringify(data.state), params }, '', newUrl);
+
+    window.scrollTo(0, 0);
+  }
+
+  return {
+    data,
+    params: merge(location.query, params),
+    pathname: newPath,
+  };
+};
+
+const findState = ({ route: state, params, data: { replace, back } }, routes, location, routesMap): any =>
+  Maybe.of(getStateKey(state, routesMap))
+    .map(ifElse(
+      path(['data', 'redirect']),
+      ({ data: { redirect } }) => findState(
+        { route: redirect, params, data: { replace: true, back } },
+        routes,
+        location,
+        routesMap
+      ),
+      ({ data, urlMatcher }) => changeState({
+        data,
+        matcher: urlMatcher,
+        params,
+        historyChange: back ? 'back' : replace ? 'replace' : 'push',
+        location,
+      }),
+    )).value();
+
+/**
+ *  @param path    the current url in the browser
+ *  @param routes  all of the possible route states
+ *
+ *  @returns state
+ *
+ *  derieves a state from given path
+ *
+ */
+const navigate = (location: Location, routes, routesMap): any => {
+  const match = matchUrl(routes, location.path, location.query);
+
+  if (match.data.redirect && !match.data.data.redirectOnFailure) {
+    return Maybe.of(findState({ route: match.data.redirect, params: match.params, data: match.data.data },
+                              routes,
+                              location,
+                              routesMap,
+    )).map(pick(['data', 'params', 'pathname'])).value();
+  }
+
+  return match;
+
+};
+
+const checkLocation = ({ current, location, e, path }) => {
+  const newLocation = exists(e.state) ? e.state : matchUrl(current.routes, path, location.query);
+
+  return equals((current.location || location), newLocation) ? null : newLocation;
+};
+
+const handler = (e: PopStateEvent, current, path, location, routesMap) =>
+  Maybe.of(checkLocation({ current, location, e, path }))
+  .map(({ state, params, data: matchedState }) => {
+    if (matchedState && matchedState.state) {
+      return { data: matchedState, params, pathname: path };
+    }
+
+    const { data, urlMatcher } = getStateKey(JSON.parse(state), routesMap);
+
+    return { data, params, pathname: urlMatcher.format(merge(params, location.query)) };
+  }).value();
+
+export const splitQuery = pipe(
+  replaceStr(/^\?/, ''),
+  split('&'),
+  map(split('=')),
+  reduce((acc, [key, val]) => ({ ...acc, [key]: val }), {}),
+);
+
+const checkCurrent = (current, config) => current.location &&
+  !equals(current.location.state, config.current.state);
+
+// tslint:disable:max-func-body-length
+export default new Map<MessageConstructor, EffectHandler>([
+  [LoadToken, ({ key, result, error }, dispatch, execContext: ExecContext<any>) => {
```

Comment:
Should probably in another pass pull out LoadToken from the Router as that's more axiom specific